### ISSUE TITLE
body:velocity:orbit:mag is always 0

### ISSUE THREAD

**Issue by GER-Space** [KSP-KOS/KOS#1374]
body:velocity:orbit:mag returns always 0. 

Body:orbit:veocity:orbit:mag returns the right value. 

What is also unexpected is: <orbitable>:velocity:orbit:mag Switches to the right value, when I'm no longer in the SOI of the planet/moon. 

example:

Kerbin:velocity:orbit:mag = 0, when I'm around kerbin 
but has the right value when I'm on duna.

**Comment by hvacengi** [comment]
I suspect that it would also be correct while you are in orbit around the sun (in transit to duna).  Orbital velocity depends on the frame of reference you're looking at.  In the Kerbin SoI, Kerbin itself does have a zero magnitude orbital velocity (since that velocity would be relative to itself.  But that's still a little inconsistent if the orbital velocity is reported in the Sun frame of reference when the ship is in the duna frame of reference.  It's also inconsistent with the implementation of `positionat()` which will return the body's position based on it's orbit about the parent.

I'll do a little looking and see if I can find the cause, and think about the implications of modifying it.

**Comment by Dunbaratu** [comment]
This is correct but confusing behavior, *assuming* the body you're talking about is the ship's current SOI body.  If this is happening for OTHER bodies, then there's a problem.  Your problem description didn't specify if the body in question was the one you are currently orbiting.

`body:velocity` means "the body's velocity in the current SHIP-RAW system", which is defined as one in which the ship's current SOI body is the reference for velocities, but the ship's current CoM is the reference for positions.  When you append `:orbit` to `body:velocity`, you just mean orbital as opposed to surface velocity.  (In the case of the current SOI body, surface and orbital velocity should come out the same because the point is at radius length zero so the planet's rotation doesn't make any difference between the two.)

Whereas, `body:orbit` means the the body's orbit that it is making around It's parent (the sun for example).

**Comment by GER-Space** [comment]
Yes the Body I'm curently orbiting is returning 0 for ist velocity. even when I'm referring it by its Name: so
KERBIN:VELOCITY:ORBIT is 0 when I'm around Kerbin and has a value when I'm anywhere else. That was not what I expected. 
Is there there any use für the local reference frame velocity? if not, than you could replace the orbitable:velocity structure with a internal reference to the orbitable:orbit:velicity structure.

**Comment by Dunbaratu** [comment]
It's not as easy as that.  What you're seeing is exposing the way the underlying KSP API is actually doing it.  The orbit's velocity is more akin to the 'on rails' Kepler ellipse - style calculation, rather than the Newtonian linear motion calculation.  In other words an orbit:velocity essentially means  "Translate this orbits' data about its current elliptical parameters and its current true anomaly into a calculation about what it's velocity has to be right now."  That style of thinking only pays attention to its local movement around its own SOI.

Ages ago, I had thought I'd want to automatically deal with reference frames for people by having every vector be aware of its own reference frame, so if you tried to, say, perform vector addition between two vectors in different reference frames, they'd know that they are in different frames and map one to the other's frame first.  But I ended up having to abandon that as a pipe dream once I realized people were writing scripts that built their own vectors from their own rotation calculations, and thus the ability to correctly track exactly which reference frame any arbitrary vector is in, without ever getting it wrong, would be unlikely.

It's one of those design sitiations where you don't want to automate something if your automation is only right 95% of the time and by automating it you've robbed your users of the ability to handle those other 5% of the cases.  i.e. I don't want to become Apple, with their infuriating mentality of "Most of our users don't need to do foo and wouldn't even understand how to do foo, therefore none of them will be allowed to do foo.  If you're one of those who do need to do foo, then you are engaging in wrongthink."

**Comment by GER-Space** [comment]
I looked into the Code:

src/kOS/Suffixed/BodyTarget.cs
Line 43
```
if (parent == null) 
```

shouldn't it read 
```
if (parent != null) 
```

**Comment by hvacengi** [comment]
Returning an body's orbital velocity of zero when within that body's sphere of influence is inconsistent with the implementation in other sections of kOS.

For starters, it means that the reference frame changes according to the ship's location.  An orbital velocity should always reference the center point of it's orbit, unless it is the center point of the orbit (for example the sun does not have an orbit in stock KSP).  This is also inconsitent with the implementation of `velocityat` and `positionat` both of which will show the progression of the parent body around it's orbit (see issue #1304 for part of that discussion).  Notably, while in orbit around Kerbin `print velocityat(kerbin, time:seconds):orbit:mag.` will return approximately "9284.5".  Likewise, `print kerbin:orbit:velocity:orbit:mag.` will return "9284.5".  But `print kerbin:velocity:orbit:mag.` returns "0".

As a further comparison, if you compare the orbital velocity while in another sphere of influence, the value returned by `kerbin:velocity:orbit` is different still.  In one test around the mun I saw 524 (the Mun's orbital velocity), while around Duna I saw 14341.

It is worth noting that if you hover over Kerbin while zoomed out on the map view, the velocity displayed is "9284.5" regardless of what sphere of influence you are currently in.

That all being said, @Dunbaratu is correct, this implementation is how KSP handles `GetVel`.  Essentially it adds up the "frame" velocity of the parent tree of each object, and then subtracts the "frame" velocity of the active vessel's reference body.  So a ship in orbit around the mun has it's orbital velocity, plus the orbital velocity of the mun, plus the orbital velocity of Kerbin, minus the orbital velocity of Kerbin.  And Duna's velocity would be reported as Duna's velocity minus Kerbin's velocity, if you're orbiting Kerbin.

I've shown a few ways to work around this problem, but we don't necessarily have to make the same mistake that we feel Squad has made.  Especially since they correct it the only place that I can find where KSP actually tells you a body's orbital velocity (the map view).

@GER-Space, the file/line you referenced above actually is "correct".  It actually is the method used by `velocityat`, not the one used by `body:velocity`.  That method will return the orbital velocity of the body that your ship is currently orbiting if you try `velocityat(sun, time:seconds)`.  The applicable method is in Utils.cs, line 32 "KOSExtensionGetObtVelocity", which returns `body.orbit.GetVel()` for the orbital velocity so long as the body contains an orbit.

**Comment by GER-Space** [comment]
Some more testing:

I have a ship on duna:

print body:velocity:orbit:mag.
0

// comsat2 is a vessel around kerbin
Set target to Vessel("comsat2").
print target:velocity:orbit:mag.
1485.3 
print target:orbit:velocity:orbit:mag.
1485.3 

// a vessel around mun
Set target to Vessel("Mun Comsat 1").
print target:velocity:orbit:mag.
345.5
print target:orbit:velocity:orbit:mag.
345.5 

// ok now for a body
Set target to mun.
print target:velocity:orbit:mag.
Result: 11771.5 
print target:orbit:velocity:orbit:mag.
Result: 542.5



when the target is a vessel, the velocity is always the orbital velocity of the vessel around the parent body.


When the target is a body I get the relative velocity to my SoI. 
As a read in the Code bodies are always on rails and to make it consitent with vessels I think it would be ok to use the similar logic of the orbit call:

src/kOS/Suffixed/BodyTarget.cs  
```
  override public OrbitableVelocity GetVelocities()
        {
           return  GetVelocitiesAtUT( new TimeSpan(Planetarium.GetUniversalTime() );
        }

```

**Comment by Dunbaratu** [comment]
@hvacengi - what you are asking for breaks the ability to add vectors nicely.  The reason for `orbitable:velocity` being always relative to the ship's current SOI is to that when you do myvelocity - othervelocity you always get the relative motion of you compared to that object.  If the meaning of velocity changed depending on where othervelocity was located rather than based on where YOU are located, you couldn't do that.

You have an inconsistency no matter what you do.  The question is which one is worse?

1. `ship:velocity` and `somethingelse:velocity` aren't always in the same reference frame as each other (it depends on what their SOI's are), thus you need a bunch of logic to catch that case before trying to do any vector math between them.

2. `somethingelse:velocity:orbit` and `somethingelse:orbit:velocity` aren't in the same reference frame as each other.  The first is always in your ship's own SOI ref frame, and the second is always in `somethingelse`'s SOI ref frame.

Between those two, I do NOT agree that 1 is somehow less of an inconsistency problem than 2 is.  2 (the way it works now) means the inconsistency is differentiated *linguistically*.  You know which was meant based on which way you traverse the objects.  "Is it somethingelse's velocity in my own reference frame, (which is what kOS always does when you get a velocity suffix directly from a thing), or is it somethingelse's velocity in its own orbit's reference frame (which is what kOS always does when you get the orbit suffix of the somethingelse first, and THEN get that's velocity)."

The way you propose makes it situationally dependant whether or not `somethingelse:velocity` is in the same reference frame your own velocity is or not.

**Comment by GER-Space** [comment]
It is more worse than that: 

In the moment Vessel:velocity is always returned as type 1
And body:velocity is returned as type 2.

So a vessel around kerbin is returning everywhere its own local velocity, even when I query is from duna and the Mun is returning the velocity relative to my current soi.

**Comment by Dunbaratu** [comment]
> In the moment Vessel:velocity is always returned as type 1
> And body:velocity is returned as type 2.
That description of the problem doesn't communicate anything I can decipher, because what I wrote as type 1 was not a single reference frame, but a description of how two reference frames differ.  And what I wrote as type 2 was the same thing - a description of how two reference frames differ, not a description of a single reference frame.  So mentioning one vector, and saying it has a reference frame like type 1 or like type 2 doesn't really make it clear what you mean.  Both kinds of reference frame were mentioned in both of those two descriptions.

**Comment by GER-Space** [comment]
Oh sorry,

I misread your post:

VESSEL:VELOCITY is returning always the velocity of the vessels Reference Frame.

BODY:VELOCITY is returning the velocity of my reference Frame. 

Example:
mysat is a satelite orbiting kerbin with at the same distance as the mun, but only 180° behind.

VESSEL("mysat"):VELOCITY:ORBIT:MAG --> always 524 no matter from where i query it. (Sun, Duna, Mun, etc)

MUN:VELOCITY:ORBIT:MAG --> depends where I am. (0, when I'm around mun, 524 around kerbin and 11000 when at Duna).

**Comment by Dunbaratu** [comment]
> VESSEL("mysat"):VELOCITY:ORBIT:MAG --> always 524 no matter from where i query it. (Sun, Duna, Mun, etc)
> 
> MUN:VELOCITY:ORBIT:MAG --> depends where I am. (0, when I'm around mun, 524 around kerbin and 11000 when at Duna).

Okay, that clears up what you meant. 
Yes if that's happening then there's a problem.  All `thing:velocity` queries were supposed to be in the current ship's SHIP-RAW frame, so they are the same as each other and can be added.  But all `thing:orbit:velocity` queries were supposed to be relative to thing's SOI because they were descriptions of thing's orbit purely in a one-body mindset - just talking about the ellipse of that one orbit.  At least that was how I tried doing it back when I refactored this stuff over a year ago with erendrake.  I may have missed a spot.

**Comment by Dunbaratu** [comment]
The problem appears to be that vesseltarget's velocity is being returned relative to *it's own* SOI, rather than relative to the ship's SOI.  When vesseltarget == ship, that's fine.  But when vesseltarget is some other ship than the SHIP, and the other vessel is in a different SOI, it's using that vessel's SOI rather than the current vessel's SOI to return the answer.

**Comment by Dunbaratu** [comment]
Punting to v1.1.1 because fixing it also means doing extensive testing of the prediction functions to make sure they match and agree, and we're getting near a deadline on v1.1.0 (1.1.0 needs to get out before SQUAD releases KSP 1.3 officially.  (It's been in  pre-release beta for over a month now so it could come any day now.)

**Comment by Dunbaratu** [comment]
### Read here for actual description of the problem:  This is the tl;dr version of the thread above:

Changed the title, because the title described *correct* behaviour as if it was the problem.  It's correct for ``body:velocity:orbit:mag`` to always be zero, because all velocities are returned in the ship's SOI body's reference frame, and ``body:velocity`` is always going to be the ship's SOI body's velocity relative to *itself*.

In the long discussion thread above, there still was a problem though (which is why I'm not closing the issue, just re-describing it).  The real problem is that:
```
vessel("some other vessel not the CPU vessel"):velocity:orbit:mag
```
is giving the same answer no matter where the CPU vessel currently is, and THAT is wrong.  What it *should* be doing is ignoring which SOI body the "some other vessel" is in, and instead phrasing the answer in terms of which SOI the CPU vessel is in.  What it's actually doing is phrasing the answer in terms of which SOI body the "some other vessel" is in.

Example: CPU vessel is in orbit around Kerbin.  OtherVessel is in orbit around Mun.   ``OtherVessel:velocity:orbit:mag`` is giving the velocity of OtherVessel around the Mun, when it should be giving its velocity around Kerbin because that's where the CPU vessel is.
